### PR TITLE
pin cmd: stream recursive pins

### PR DESCRIPTION
**Pull request by MichaelMure**
Add a --stream flag to stream the results instead of
accumulating the final result in memory.

This is a rework of https://github.com/ipfs/go-ipfs/pull/5005

### CODE REVIEW

**Comment by MichaelMure** [comment]
Notes:
- pins are streamed and start being emitted immediately
- direct and recursive are streamed super fast first, around 76k pins/s on my test machine
- indirect pins are still streamed very slowly, around 5 pins/s, most likely due to no parallelization in go-merkledag: https://github.com/ipfs/go-merkledag/blob/master/merkledag.go#L287

File: core/commands/pin.go [hunk 1]
```diff
@@ -492,80 +508,127 @@ var verifyPinCmd = &cmds.Command{
...
 type RefKeyList struct {
-	Keys map[string]RefKeyObject
+	Keys map[string]RefObject
 }
 
-func pinLsKeys(ctx context.Context, args []string, typeStr string, n *core.IpfsNode, api coreiface.CoreAPI) (map[cid.Cid]RefKeyObject, error) {
+// Pin ls needs to output two different type depending on if it's streamed or not.
+// We use this to bypass the cmds lib refusing to have interface{}
+type PinLsOutputWrapper struct {
+	RefKeyList
```
Comment (Stebalien) [inline core/commands/pin.go]: We're going to have to use "omitempty" for full backwards compatibility.

**Comment by MichaelMure** [comment]
> This will need to correctly handle CID encoding (introduced since #5005).

Any pointer/example about this ? I don't really get what you are talking about.

File: core/commands/pin.go [hunk 2]
```diff
@@ -337,34 +337,50 @@ Example:
...
-			keys, err = pinLsAll(req.Context, typeStr, n)
+			err = pinLsAll(req.Context, typeStr, n, emit)
 		}
 		if err != nil {
 			return err
 		}
 
-		refKeys := make(map[string]RefKeyObject, len(keys))
-		for k, v := range keys {
-			refKeys[enc.Encode(k)] = v
```
Comment (Stebalien) [inline core/commands/pin.go]: Encoding code.

File: core/commands/pin.go [hunk 3]
```diff
@@ -337,34 +337,50 @@ Example:
 			return err
 		}
 
-		enc, err := cmdenv.GetCidEncoder(req)
```
Comment (Stebalien) [inline core/commands/pin.go]: Other encoding code.

**Comment by MichaelMure** [comment]
A few improvements:
- CID are encoded properly
- better struct naming (they were named after the `refs` command but are only used by `pin/ls` now)
- struct field are properly JSON omitempty-ed for backward compat
- output struct are documented to make codeclimate happy

**Comment by MichaelMure** [comment]
@lanzafame not sure how I would do a go benchmark for this. This PR doesn't fundamentally change anything perf wise, it just start to output the result as it comes. The total run time will be similar.

**Comment by MichaelMure** [comment]
More improvements:
- fixed a bug where indirect pins were reported as "all"
- output the recursive pins first as they don't require walking the graph
- use `EnumerateChildrenAsync` instead of `EnumerateChildren` for faster/parallel walking of the graph

```
root@ip-10-0-25-166:/tmp# timeout -k 1 59s ./ipfsseq pin ls --stream --type=indirect > indirectseq
12:50:51.021 ERROR       core: failure on stop:  context canceled builder.go:47
Killed
root@ip-10-0-25-166:/tmp# wc -l indirectseq
786 indirectseq
root@ip-10-0-25-166:/tmp# timeout -k 1 59s ./ipfsasync pin ls --stream --type=indirect > indirectasync
12:52:52.668 ERROR       core: failure on stop:  context canceled builder.go:47
Error: canceled
root@ip-10-0-25-166:/tmp# wc -l indirectasync
2645 indirectasync
```

For indirect pins:
Sequential walking: 13 pins/s
Parallel walking: 44 pins/s

--> ~3.4x faster !

It also looks like this change to use parallel graph walking might be used elsewhere. A quick search show that at least the GC might benefit from that.

**Comment by MichaelMure** [comment]
Test are now failing due to using `EnumerateChildrenAsync`. If `EnumerateChildren` is used instead, tests are ok (at least as far as I can tell, I have some unrelated failing on my machine).

**Comment by Stebalien** [comment]
>  is used instead, tests are ok (at least as far as I can tell, I have some unrelated failing on my machine).

We probably need to sort in the tests.

**Comment by MichaelMure** [comment]
FYI, I removed the commit that just change `EnumerateChildren` to `EnumerateChildrenAsync`. As you can see, the CI is now okay-ish. If you have insights on the remaining failure and why using `EnumerateChildrenAsync` could be problematic, I'm interested.

**Comment by Stebalien** [comment]
I'm _guessing_ that the traversal order is random, causing the comparison against the expected pin list to fail. We need to sort both before comparing.

**Comment by MichaelMure** [comment]
I checked that, but the tests look ok to me:
https://github.com/ipfs/interop/blob/master/test/pin.js#L89-L90
https://github.com/ipfs/interop/blob/master/test/pin.js#L155-L157

**Comment by Stebalien** [comment]
So... this is a bug. `EnumerateChildren` enumerates children. `EnumerateChildrenAsync` _also_ enumerates the root CID. I believe we should make it operate like `EnumerateChildren` but we're going to have to carefully make sure we don't break something in the process.

**Comment by MichaelMure** [comment]
Hoooo, that explains a lot. One problem I had after adding the recursive pins before the indirect pins to the set and using `EnumerateChildrenAsync` was that indirect pins would not be listed anymore. It was because the root cid was returned first by `EnumerateChildrenAsync`, considered as already enumerated and the whole subgraph was pruned.

**Comment by Stebalien** [comment]
Fix in https://github.com/ipfs/go-ipfs/pull/6499. Well, not really a _fix_, I just renamed these functions so that the names/documentation match up with what they do...

**Comment by lanzafame** [comment]
> @lanzafame not sure how I would do a go benchmark for this. This PR doesn't fundamentally change anything perf wise, it just start to output the result as it comes. The total run time will be similar.

With a benchmark, you can run `go test -run none -bencmark . -benchmem` and you will get number of allocations etc.

**Comment by Stebalien** [comment]
@alanshaw this PR adds a `--stream` (short `-s`) option to `ipfs pin ls` that returns of pins as a stream of :

```json
{"Cid": "Qm1", "Type": "direct"}
{"Cid": "Qm2", "Type": "recursive"}
...
```

Instead of a single:

```json
{
  "Keys": {
    "Qm1": "direct",
    "Qm2": "recursive",
  }
}
```

This is completely backwards compatible but is likely something js-ipfs will want to implement. Thoughts? Objections?

**Comment by alanshaw** [comment]
Yes please. I want this. No objections.

**Comment by Stebalien** [comment]
:rocket:!

Thanks @MichaelMure, @lanzafame, and @alanshaw!

**Comment by MichaelMure** [comment]
Ha, I was waiting for https://github.com/ipfs/go-ipfs/pull/6499#issuecomment-510783458 to get resolved so I could have the neat 3.4x perf improvement with the parallel walking, but ok !

**Comment by Stebalien** [comment]
Test case in https://github.com/ipfs/go-ipfs/pull/6504.

Note: IMO, a benchmark isn't really useful here as this _just_ streams the results, it doesn't reduce overall memory (still need to deduplicate results by storing them in a map).

**Comment by Stebalien** [comment]
> Ha, I was waiting for #6499 (comment) to get resolved so I could have the neat 3.4x perf improvement with the parallel walking, but ok !

Let's do that in a followup. No reason to block this on a different change.

**Pull request merged**